arangod/Scheduler/Scheduler.cpp

```diff
@@ -370,16 +455,30 @@ bool Scheduler::pushToFifo(size_t fifo, std::function<void()> const& callback) {
   return true;
 }
 
-bool Scheduler::popFifo(size_t fifo) {
-  int64_t p = fifo - 1;
-  TRI_ASSERT(0 <= p && p < NUMBER_FIFOS);
+bool Scheduler::popFifo(int64_t fifo) {
+  TRI_ASSERT(0 <= fifo && fifo < NUMBER_FIFOS);
+
+  if (fifo == FIFO8 && _queuedV8 >= _maxQueuedV8) {
+    return false;
+  }
 
+  size_t p = static_cast<size_t>(fifo);
   FifoJob* job = nullptr;
+
   bool ok = _fifos[p]->pop(job) && job != nullptr;
 
   if (ok) {
-    post(job->_callback);
-    delete job;
+    auto guard = scopeGuard([job]() {
+      if (job) {
+        delete job;
+      }
+    });
```

Comment:
Isn't this exactly the logic of a `unique_ptr`?

Reply:
yes, could be used as well